ludwig/distributed/deepspeed.py

```diff
@@ -57,6 +59,7 @@ def __init__(
 
         super().__init__(**kwargs)
         self.zero_optimization = zero_optimization or DEFAULT_ZERO_OPTIMIZATION
+        self.zero_optimization_stage = self.zero_optimization.get("stage", 3)
```

Comment:
Why default to 3?

Reply:
It's because in the worst case, it makes sense to assume that the model does not fit into memory and we need to do model parallel + data parallel training since the model will not fit into a single GPU. It also assumes no quantization based fine-tuning, which I feel like is a far assumption to make. All of this is to say that we want a config as simple as this "to just work"

```yaml
input_features: ...
output_features: ...
trainer:
    type: finetune
backend:
    type: ray
    trainer:
        strategy:
            type: deepspeed
```

to have the highest chance of succeeding irrespective of LLM model size. This is the worst case scenario that does full fine-tuning with an adapter or without quantization. All of those only reduce the model size, but this is the most useful configuration to default to assuming none of those are set in the config for LLM fine-tuning.

Let me know if this makes sense. I'll also add a comment in the DeepSpeed class explaining why we default to stage 3.